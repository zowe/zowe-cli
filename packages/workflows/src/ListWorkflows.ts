/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ZosmfRestClient, nozOSMFVersion } from "@zowe/core-for-zowe-sdk";
import { WorkflowValidator } from "./WorkflowValidator";
import { AbstractSession, ImperativeError } from "@zowe/imperative";
import { WorkflowConstants, wrongString, noWorkflowName } from "./WorkflowConstants";
import { IActiveWorkflows } from "./doc/IActiveWorkflows";
import { IGetWorkflowsOptions } from "./doc/IGetWorkflowsOptions";

/**
 * Get list of workflows from registry.
 * @export
 * @class ListWorkflows
 */
export class ListWorkflows {
    /**
     * This operation returns list of workflows.
     * Parameters are optional,request can include one or more parameters to filter the results.
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} zOSMFVersion - the URI path that identifies the version of the provisioning service.
     * @param {string} workflowName - the URI path with optional parameter for listing filtered workflows.
     * @param {string} category - the URI path with optional parameter for listing filtered workflows.
     * @param {string} system - the URI path with optional parameter for listing filtered workflows.
     * @param {string} owner - the URI path with optional parameter for listing filtered workflows.
     * @param {string} vendor - the URI path with optional parameter for listing filtered workflows.
     * @param {string} statusName - the URI path with optional parameter for listing filtered workflows.
     * @returns {string} z/OSMF response object
     * @memberof ListWorkflows
     * @deprecated
     */
    public static async listWorkflows(session: AbstractSession, zOSMFVersion = WorkflowConstants.ZOSMF_VERSION,
        workflowName?: string, category?: string, system?: string,
        owner?: string, vendor?: string, statusName?: string ) {
        return this.getWorkflows(session, {zOSMFVersion, workflowName, category, system, owner, vendor, statusName});
    }

    /**
     * This operation returns list of workflows.
     * Parameters are optional,request can include one or more parameters to filter the results.
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {IGetWorkflowsOptions} options - Options to filter the request
     * @returns {string} z/OSMF response object
     * @memberof ListWorkflows
     */
    public static async getWorkflows(session: AbstractSession, options: IGetWorkflowsOptions = {}) {
        WorkflowValidator.validateSession(session);
        options = {
            ...options,
            // zOSMFVersion by design was only checking for undefined. To prevent a breaking change we have to do the same ??
            // Added another "system" test for this
            zOSMFVersion: options.zOSMFVersion !== undefined ? options.zOSMFVersion : WorkflowConstants.ZOSMF_VERSION,
            workflowName: options.workflowName ? encodeURIComponent(options.workflowName) : null
        };
        WorkflowValidator.validateNotEmptyString(options.zOSMFVersion, nozOSMFVersion.message);
        const resourcesQuery: string = ListWorkflows.getResourceQuery(options);
        return ZosmfRestClient.getExpectJSON(session, resourcesQuery);
    }

    /**
     * This operation Builds URI path from provided parameters.
     * @param {string} zOSMFVersion - the URI path that identifies the version of the provisioning service.
     * @param {string} params - The array with URI path with filters for listing filtered workflows.
     * @returns {string} URI path for the REST call.
     * @memberof ListWorkflows
     * @deprecated
     */
    public static getResourcesQuery(zOSMFVersion: string, params: Array <{key: string, value: string}>) {
        let query: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/${WorkflowConstants.WORKFLOW_RESOURCE}`;
        let sign = "?";
        params.forEach((element) => {
            if (element.value) {
                // Validate if parameter value does not contains ? or &
                WorkflowValidator.validateParameter(element.value, wrongString.message);
                query += sign + `${element.key}=${element.value}`;
                sign = "&";
            }
        });
        return query;
    }

    /**
     * This operation Builds URI path from provided parameters.
     * @param {IGetWorkflowsOptions} params - The array with URI path with filters for listing filtered workflows.
     * @returns {string} URI path for the REST call.
     * @memberof ListWorkflows
     */
    public static getResourceQuery(params: IGetWorkflowsOptions) {
        let query: string = `${WorkflowConstants.RESOURCE}/${params.zOSMFVersion}/${WorkflowConstants.WORKFLOW_RESOURCE}`;
        let sign = "?";
        for (const key in params) {
            if (key === "zOSMFVersion") continue;
            if (params[key]) {
                // Validate if parameter value does not contains ? or &
                WorkflowValidator.validateParameter(params[key], wrongString.message);
                query += sign + `${key}=${params[key]}`;
                sign = "&";
            }
        }
        return query;
    }

    /**
     * This operation is used to return a worflow-key by given workflow name.
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} workflowName - workflow name by which to list workflows
     * @param {string} zOSMFVersion - identifies the version of the provisioning service.
     * @returns {Promise<string> | null} - Promise with string containing wf key, or null if none was found
     * @throws {ImperativeError}
     * @memberof ListWorkflows
     */
    public static async getWfKey(session: AbstractSession, workflowName: string,
        zOSMFVersion = WorkflowConstants.ZOSMF_VERSION): Promise<string> | null{

        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
        WorkflowValidator.validateNotEmptyString(workflowName, noWorkflowName.message);

        const result: IActiveWorkflows = await this.getWorkflows(session, {workflowName});

        // Check if there was more than one workflows found
        if (result.workflows.length > 1){
            throw new ImperativeError({
                msg: `More than one workflows found with name ` + workflowName
            });
        }
        return result.workflows.length !== 0 ? result.workflows[0].workflowKey : null;
    }
}
