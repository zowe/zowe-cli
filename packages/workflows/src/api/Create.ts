/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/


import { AbstractSession, Headers, ImperativeError } from "@brightside/imperative";
import { ZosmfRestClient } from "../../../rest";
import {
    WorkflowConstants,
    nozOSMFVersion,
    noWorkflowName,
    noWorkflowDefinitionFile,
    noSystemName,
    noOwner,
    wrongOwner, wrongPath
} from "./WorkflowConstants";
import { WorkflowValidator } from "./WorkflowValidator";
import { isNullOrUndefined } from "util";
import { ICreateWorkflow, accessT } from "./doc/ICreateWorkflow";
import { ICreatedWorkflow } from "./doc/ICreatedWorkflow";
import { IVariable } from "./doc/IVariables";

/**
 * Class to handle creation of zOSMF workflow instance
 */
export class CreateWorkflow{

    /**
     * copied from ProvisioningService.ts
     * Parsers text with properties in key1=val1,key2=val2 format and returns IInputProperty[]
     * @param {string} propertiesText - required runtime property objects passed as a string.
     * @returns {IPropertiesInput[]} array of properties, @see {IPropertiesInput}
     * @memberof ProvisioningService
     */
    public static parseProperties(propertiesText: string): IVariable[] {
        if (propertiesText === "") {
            return [];
        }
        return propertiesText.split(",").map((property) => {
            const tempArray = property.split("=");
            if (tempArray.length === 2 && tempArray[0].length > 0) {
                return {name: tempArray[0].trim(), value: tempArray[1].trim()};
            } else {
                throw new ImperativeError({msg: `Incorrect properties format: ${propertiesText}`});
            }
        });
    }
    /**
     * Create a zOSMF workflow instance
     * @param {AbstractSession} session                     - z/OSMF connection info
     * @param {string} WorkflowName                         - Name of the workflow that will be created
     * @param {string} WorkflowDefinitionFile               - Full path to USS file or DATASET/MEMBER with zml
     * @param {string} systemName                           - System where the workflow will run
     * @param {string} Owner                                - User ID of the workflow owner.
     * @param {string} VariableInputFile                    - Properties file with pre-specify values for workflow variables
     * @param {string} Variables                            - A list of one or more variables for the workflow.
     * @param {boolean} AssignToOwner                       - Indicates whether the workflow steps are assigned to the workflow owner
     * @param {accessT} AccessType                          - Specifies the access type for the workflow. Public, Restricted or Private.
     * @param {boolean} DeleteCompletedJobs                 - Specifies whether the job is deleted from the JES spool after it completes successfully.
     * @param {string} zOSMFVersion                         - Identifies the version of the zOSMF workflow service.
     * @returns {Promise<ICreatedWorkflow>}
     */
    public static createWorkflow(session: AbstractSession, WorkflowName: string, WorkflowDefinitionFile: string,
                                 systemName: string, Owner: string, VariableInputFile?: string, Variables?: string,
                                 AssignToOwner?: boolean, AccessType?: accessT, DeleteCompletedJobs?: boolean,
                                 zOSMFVersion = WorkflowConstants.ZOSMF_VERSION,
                                    // add job statement, account info, comments and resolveGlobalConflictByUsing,
                                    ): Promise<ICreatedWorkflow> {

        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
        WorkflowValidator.validateNotEmptyString(WorkflowName, noWorkflowName.message);
        WorkflowValidator.validateNotEmptyString(WorkflowDefinitionFile, noWorkflowDefinitionFile.message);
        WorkflowValidator.validatePath(WorkflowDefinitionFile, wrongPath.message);
        WorkflowValidator.validateNotEmptyString(systemName, noSystemName.message);
        WorkflowValidator.validateNotEmptyString(Owner, noOwner.message);
        WorkflowValidator.validateOwner(Owner, wrongOwner.message);

        const data: ICreateWorkflow = {
            workflowName: WorkflowName,
            workflowDefinitionFile: WorkflowDefinitionFile,
            system: systemName,
            owner: Owner,
            assignToOwner: AssignToOwner,
            accessType: AccessType,
            deleteCompletedJobs: DeleteCompletedJobs,
        };
        if (!isNullOrUndefined(VariableInputFile)){
            WorkflowValidator.validatePath(VariableInputFile, wrongPath.message);
            data.variableInputFile = VariableInputFile;
        }
        if (!isNullOrUndefined(Variables)){
            data.variables = this.parseProperties(Variables);
        }
        if (isNullOrUndefined(AssignToOwner)){
            data.assignToOwner = true;
        }
        if (isNullOrUndefined(AccessType)){
            data.accessType = "Public";
        }
        if (isNullOrUndefined(DeleteCompletedJobs)){
            data.deleteCompletedJobs = false;
        }

        const resourcesQuery: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/${WorkflowConstants.WORKFLOW_RESOURCE}`;

        return ZosmfRestClient.postExpectJSON<ICreatedWorkflow>(session, resourcesQuery, [Headers.APPLICATION_JSON], data);
    }
}
