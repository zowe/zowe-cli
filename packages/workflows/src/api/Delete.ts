/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/


import { AbstractSession } from "@brightside/imperative";
import { ZosmfRestClient } from "../../../rest";
import { WorkflowConstants, nozOSMFVersion, noWorkflowKey } from "./WorkflowConstants";
import { WorkflowValidator } from "./WorkflowValidator";

/**
 * Class to handle deletion of zOSMF workflow instance
 */
export class DeleteWorkflow {

    /**
     * Delete a workflow instance
     * @param {AbstractSession} session                     - z/OSMF connection info
     * @param {string} workflowKey                          - Unique identifier of the workflow instance.
     * @param {string} zOSMFVersion                         - Identifies the version of the zOSMF workflow service.
     * @returns {string}
     */
 public static async deleteWorkflow(session: AbstractSession, workflowKey: string,
                                    zOSMFVersion = WorkflowConstants.ZOSMF_VERSION){
    WorkflowValidator.validateSession(session);
    WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
    WorkflowValidator.validateNotEmptyString(workflowKey, noWorkflowKey.message);
    let resourcesQuery: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/`;
    resourcesQuery += `${WorkflowConstants.WORKFLOW_RESOURCE}/${workflowKey}`;

    return ZosmfRestClient.deleteExpectString(session, resourcesQuery, []);
 }
}
