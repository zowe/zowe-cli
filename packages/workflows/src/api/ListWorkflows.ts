/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ZosmfRestClient } from "../../../rest";
import { WorkflowValidator } from "./WorkflowValidator";
import { AbstractSession, ImperativeError } from "@zowe/imperative";
import { WorkflowConstants, nozOSMFVersion, wrongString, noWorkflowName } from "./WorkflowConstants";
import { IActiveWorkflows } from "./doc/IActiveWorkflows";

/**
 * Get list of workflows from registry.
 * @export
 * @class ListWorkflows
 */
export class ListWorkflows {
    /**
     * This operation returns list of workflows.
     * Parameters are optional,request can include one or more parameters to filter the results.
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} zOSMFVersion - the URI path that identifies the version of the provisioning service.
     * @param {string} workflowName - the URI path with optional parameter for listing filtered workflows.
     * @param {string} category - the URI path with optional parameter for listing filtered workflows.
     * @param {string} system - the URI path with optional parameter for listing filtered workflows.
     * @param {string} owner - the URI path with optional parameter for listing filtered workflows.
     * @param {string} vendor - the URI path with optional parameter for listing filtered workflows.
     * @param {string} statusName - the URI path with optional parameter for listing filtered workflows.
     * @returns {string} z/OSMF response object
     * @memberof ListWorkflows
     */
    public static async listWorkflows(session: AbstractSession, zOSMFVersion = WorkflowConstants.ZOSMF_VERSION,
                                      workflowName?: string, category?: string, system?: string,
                                      owner?: string, vendor?: string, statusName?: string ) {
        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
        const resourcesQuery: string = ListWorkflows.getResourcesQuery(zOSMFVersion,
            [
                {key: WorkflowConstants.workflowName, value : workflowName ? encodeURIComponent(workflowName) : null},
                {key: WorkflowConstants.category, value : category},
                {key: WorkflowConstants.system, value : system},
                {key: WorkflowConstants.owner, value : owner},
                {key: WorkflowConstants.vendor, value : vendor},
                {key: WorkflowConstants.statusName, value : statusName}
            ]
        );
        return ZosmfRestClient.getExpectJSON(session, resourcesQuery);
    }

    /**
     * This operation Builds URI path from provided parameters.
     * @param {string} zOSMFVersion - the URI path that identifies the version of the provisioning service.
     * @param {string} params - The array with URI path with filters for listing filtered workflows.
     * @returns {string} URI path for the REST call.
     * @memberof ListWorkflows
     */
    public static getResourcesQuery(zOSMFVersion: string, params: Array <{key: string, value: string}>) {
        let query: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/${WorkflowConstants.WORKFLOW_RESOURCE}`;
        let sign = "?";
        params.forEach((element) => {
            if (element.value) {
                // Validate if parameter value does not contains ? or &
                WorkflowValidator.validateParameter(element.value, wrongString.message);
                query += sign + `${element.key}=${element.value}`;
                sign = "&";
            }
        });
        return query;
    }

    /**
     * This operation is used to return a worflow-key by given workflow name.
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} workflowName - workflow name by which to list workflows
     * @param {string} zOSMFVersion - identifies the version of the provisioning service.
     * @returns {Promise<string> | null} - Promise with string containing wf key, or null if none was found
     * @throws {ImperativeError}
     * @memberof ListWorkflows
     */
    public static async getWfKey(session: AbstractSession, workflowName: string,
                                 zOSMFVersion = WorkflowConstants.ZOSMF_VERSION): Promise<string> | null{

        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
        WorkflowValidator.validateNotEmptyString(workflowName, noWorkflowName.message);

        const result: IActiveWorkflows = await this.listWorkflows(session, zOSMFVersion, workflowName);

        // Check if there was more than one workflows found
        if (result.workflows.length > 1){
            throw new ImperativeError({
                msg: `More than one workflows found with name ` + workflowName
            });
        }
        return result.workflows.length !== 0 ? result.workflows[0].workflowKey : null;
    }
}
