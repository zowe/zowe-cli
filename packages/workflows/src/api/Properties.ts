/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/


import { AbstractSession, Headers } from "@brightside/imperative";
import { ZosmfRestClient } from "../../../rest";
import { WorkflowConstants, nozOSMFVersion,
        noOwner, noWorkflowKey } from "./WorkflowConstants";
import { WorkflowValidator } from "./WorkflowValidator";
import { isNullOrUndefined } from "util";
import { IWorkflowInfo } from "./doc/IWorkflowInfo";

export class PropertiesWorkflow {
    /**
     * This operation returns properties of the workflow.
     * Parameters indicators are mandatory,request can include steps and variables indicator for requested result.
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} workflowfKey - Key of workflow.
     * @param {string} zOSMFVersion - the URI path that identifies the version of the provisioning service.
     * @param {boolean} steps - Optional parameter for listing steps properties.
     * @param {boolean} variables - Optional parameter for listing variables properties.
     * @returns {Promise<IWorkflowInfo>} z/OSMF response object
     * @memberof Properties
     */
    // main method
    public static async getWorkflowProperties(session: AbstractSession, workflowKey: string, zOSMFVersion = WorkflowConstants.ZOSMF_VERSION,
                                              steps: boolean, variables: boolean): Promise<IWorkflowInfo>{
        WorkflowValidator.validateSession(session);
        WorkflowValidator.validateNotEmptyString(zOSMFVersion, nozOSMFVersion.message);
        let wfKey: string;

        WorkflowValidator.validateNotEmptyString(workflowKey, noWorkflowKey.message);
        wfKey = workflowKey;

        let resourcesQuery: string = `${WorkflowConstants.RESOURCE}/${zOSMFVersion}/`;
        resourcesQuery += `${WorkflowConstants.WORKFLOW_RESOURCE}/${wfKey}`;

        if (steps && variables){
            resourcesQuery += `?${WorkflowConstants.returnData}=${WorkflowConstants.steps},${WorkflowConstants.variables}`;

        } else if (steps)   {
            resourcesQuery += `?${WorkflowConstants.returnData}=${WorkflowConstants.steps}`;

        } else if (variables)   {
            resourcesQuery += `?${WorkflowConstants.returnData}=${WorkflowConstants.variables}`;

        }

        return ZosmfRestClient.getExpectJSON<IWorkflowInfo>(session, resourcesQuery, [Headers.APPLICATION_JSON]);
    }

}

