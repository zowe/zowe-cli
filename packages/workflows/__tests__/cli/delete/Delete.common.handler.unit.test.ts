/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { DeleteWorkflow } from "../../../src/api/Delete";
import { ListWorkflows } from "../../../src/api/ListWorkflows";
import { AbstractSession, ImperativeError } from "@zowe/imperative";


describe("Delete workflow common handler", () => {
    describe("process method", () => {
        it("should delete a workflow using workflow key", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../src/cli/delete/Delete.common.handler");
            const handler = new handlerReq.default();
            const workflowKey = "fake-workflow-key";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession: AbstractSession = null;

            // Mock the delete function
            DeleteWorkflow.deleteWorkflow = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "deleted"
                };
            });

            // Mocked function references
            const profFunc = jest.fn((args) => {
                return {
                    host: "fake",
                    port: "fake",
                    user: "fake",
                    pass: "fake",
                    auth: "fake",
                    rejectUnauthorized: "fake",
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.processCmd({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        workflowKey
                    },
                    response: {
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        }
                    },
                    profiles: {
                        get: profFunc
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeUndefined();
            expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledTimes(1);
            expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledWith(fakeSession, workflowKey);

        });
        it("should delete a workflow using workflow name", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../src/cli/delete/Delete.common.handler");
            const handler = new handlerReq.default();
            const wfArray = {workflows: [
                {workflowKey: "fake-workflow-key"},
                {workflowKey: "fake-workflow-key2"}
            ]};
            const workflowName = "fake-name";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession: AbstractSession = null;

            // Mock the delete function
            DeleteWorkflow.deleteWorkflow = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "deleted"
                };
            });

            // Mock the list function
            ListWorkflows.listWorkflows = jest.fn((session) => {
                fakeSession = session;
                return wfArray;
            });

            // Mocked function references
            const profFunc = jest.fn((args) => {
                return {
                    host: "fake",
                    port: "fake",
                    user: "fake",
                    pass: "fake",
                    auth: "fake",
                    rejectUnauthorized: "fake",
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.processCmd({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        workflowName
                    },
                    response: {
                        format: {
                            output: jest.fn((parms) => {
                                expect(parms).toMatchSnapshot();
                            })
                        },
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        }
                    },
                    profiles: {
                        get: profFunc
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeUndefined();
            expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledTimes(wfArray.workflows.length);
            wfArray.workflows.forEach((element)=>{
                expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledWith(fakeSession, element.workflowKey);
            });
        });
    });
    describe("Fail scenarios", () => {
        it("should fail if no workflow with specified name was found.", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../src/cli/delete/Delete.common.handler");
            const handler = new handlerReq.default();
            const workflowName = "fake-name";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession: AbstractSession = null;

            // Mock the delete function
            DeleteWorkflow.deleteWorkflow = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "deleted"
                };
            });

            // Mock the list function
            ListWorkflows.listWorkflows = jest.fn((session) => {
                fakeSession = session;
                return null;
            });

            // Mocked function references
            const profFunc = jest.fn((args) => {
                return {
                    host: "fake",
                    port: "fake",
                    user: "fake",
                    pass: "fake",
                    auth: "fake",
                    rejectUnauthorized: "fake",
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.processCmd({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        workflowName
                    },
                    response: {
                        format: {
                            output: jest.fn((parms) => {
                                expect(parms).toMatchSnapshot();
                            })
                        },
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        }
                    },
                    profiles: {
                        get: profFunc
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(JSON.stringify(error)).toContain("No workflows match the provided workflow name.");
            expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledTimes(0);
        });
        it("should fail if wrong workflow deletion criteria was provided.", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../src/cli/delete/Delete.common.handler");
            const handler = new handlerReq.default();
            const workflowID = "fake-id";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession: AbstractSession = null;

            // Mock the delete function
            DeleteWorkflow.deleteWorkflow = jest.fn((session) => {
                fakeSession = session;
                return {
                    success: true,
                    commandResponse: "deleted"
                };
            });

            // Mock the list function
            ListWorkflows.listWorkflows = jest.fn((session) => {
                fakeSession = session;
                return null;
            });

            // Mocked function references
            const profFunc = jest.fn((args) => {
                return {
                    host: "fake",
                    port: "fake",
                    user: "fake",
                    pass: "fake",
                    auth: "fake",
                    rejectUnauthorized: "fake",
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.processCmd({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        workflowID
                    },
                    response: {
                        format: {
                            output: jest.fn((parms) => {
                                expect(parms).toMatchSnapshot();
                            })
                        },
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        }
                    },
                    profiles: {
                        get: profFunc
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(JSON.stringify(error)).toContain(
                "Internal create error: Unable to determine the the criteria by which to run delete workflow action.");
            expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledTimes(0);
        });
        it("should fail if deletion with wf key failed", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../src/cli/delete/Delete.common.handler");
            const handler = new handlerReq.default();
            const workflowKey = "fake-wf-key";
            const deleteError: string = "failed to delete wf";

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession: AbstractSession = null;

            // Mock the delete function
            DeleteWorkflow.deleteWorkflow = jest.fn((session) => {
                fakeSession = session;
                throw new ImperativeError({msg: deleteError});
            });

            // Mocked function references
            const profFunc = jest.fn((args) => {
                return {
                    host: "fake",
                    port: "fake",
                    user: "fake",
                    pass: "fake",
                    auth: "fake",
                    rejectUnauthorized: "fake",
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.processCmd({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        workflowKey
                    },
                    response: {
                        format: {
                            output: jest.fn((parms) => {
                                expect(parms).toMatchSnapshot();
                            })
                        },
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        }
                    },
                    profiles: {
                        get: profFunc
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(JSON.stringify(error)).toContain("Delete workflow: Error: " + deleteError);
            expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledTimes(1);
        });
        it("should fail if deletion with wf name failed", async () => {
            // Require the handler and create a new instance
            const handlerReq = require("../../../src/cli/delete/Delete.common.handler");
            const handler = new handlerReq.default();
            const workflowName = "fake-wf-name";
            const deleteError: string = "failed to delete wf";
            const wfArray = {workflows: [
                {workflowKey: "fake-workflow-key"},
                {workflowKey: "fake-workflow-key2"}
            ]};

            // Vars populated by the mocked function
            let error;
            let apiMessage = "";
            let jsonObj;
            let logMessage = "";
            let fakeSession: AbstractSession = null;

            // Mock the delete function
            DeleteWorkflow.deleteWorkflow = jest.fn((session) => {
                fakeSession = session;
                throw new ImperativeError({msg: deleteError});
            });

            // Mock the list function
            ListWorkflows.listWorkflows = jest.fn((session) => {
                fakeSession = session;
                return wfArray;
            });

            // Mocked function references
            const profFunc = jest.fn((args) => {
                return {
                    host: "fake",
                    port: "fake",
                    user: "fake",
                    pass: "fake",
                    auth: "fake",
                    rejectUnauthorized: "fake",
                };
            });

            try {
                // Invoke the handler with a full set of mocked arguments and response functions
                await handler.processCmd({
                    arguments: {
                        $0: "fake",
                        _: ["fake"],
                        workflowName
                    },
                    response: {
                        format: {
                            output: jest.fn((parms) => {
                                expect(parms).toMatchSnapshot();
                            })
                        },
                        data: {
                            setMessage: jest.fn((setMsgArgs) => {
                                apiMessage = setMsgArgs;
                            }),
                            setObj: jest.fn((setObjArgs) => {
                                jsonObj = setObjArgs;
                            })
                        },
                        console: {
                            log: jest.fn((logArgs) => {
                                logMessage += "\n" + logArgs;
                            })
                        }
                    },
                    profiles: {
                        get: profFunc
                    }
                } as any);
            } catch (e) {
                error = e;
            }

            expect(error).toBeDefined();
            expect(JSON.stringify(error)).toContain("Some workflows were not deleted, please check the message above.");
            expect(DeleteWorkflow.deleteWorkflow).toHaveBeenCalledTimes(wfArray.workflows.length);
        });
    });
});
