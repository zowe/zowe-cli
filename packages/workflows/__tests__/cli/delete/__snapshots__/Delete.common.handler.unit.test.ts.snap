// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete workflow common handler Fail scenarios should fail if deletion with wf name failed 1`] = `
Object {
  "fields": Array [
    "workflowName",
    "workflowKey",
  ],
  "format": "table",
  "header": true,
  "output": Array [],
}
`;

exports[`Delete workflow common handler Fail scenarios should fail if deletion with wf name failed 2`] = `
Object {
  "fields": Array [
    "workflowName",
    "workflowKey",
  ],
  "format": "table",
  "header": true,
  "output": Array [
    Object {
      "workflowKey": "fake-workflow-key",
    },
    Object {
      "workflowKey": "fake-workflow-key2",
    },
  ],
}
`;

exports[`Delete workflow common handler process method should delete a workflow using workflow name 1`] = `
Object {
  "fields": Array [
    "workflowName",
    "workflowKey",
  ],
  "format": "table",
  "header": true,
  "output": Array [
    Object {
      "workflowKey": "fake-workflow-key",
    },
    Object {
      "workflowKey": "fake-workflow-key2",
    },
  ],
}
`;
